pkg/trigger/migration_handler.go

```diff
@@ -0,0 +1,108 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trigger
+
+import (
+	"fmt"
+	"reflect"
+
+	migrationv1alpha1 "github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/apis/migration/v1alpha1"
+	"github.com/kubernetes-sigs/kube-storage-version-migrator/pkg/controller"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func storageStateName(resource migrationv1alpha1.GroupVersionResource) string {
+	// TODO: add this rule to the CRD validation
+	// TODO: we might use ResourceID as the name in the future.
+	return resource.Resource + "." + resource.Group
+}
+
+func (mt *MigrationTrigger) markStorageStateSucceeded(resource migrationv1alpha1.GroupVersionResource) error {
+	// We will retry on any error. Migrating a resource takes a long time.
+	// It would be a pity to give up just because of an update error.
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		ss, err := mt.client.MigrationV1alpha1().StorageStates().Get(storageStateName(resource), metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			utilruntime.HandleError(err)
+			return false, nil
+		}
+		if err != nil && errors.IsNotFound(err) {
+			utilruntime.HandleError(err)
+			// if the storage state does not exist, there is not
+			// much we can do. We will leave it to the next
+			// discovery routine to create the storage state.
+			return true, nil
+		}
+		ss.Status.PersistedStorageVersionHashes = []string{ss.Status.CurrentStorageVersionHash}
+		_, err = mt.client.MigrationV1alpha1().StorageStates().Update(ss)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+func (mt *MigrationTrigger) launchMigration(resource migrationv1alpha1.GroupVersionResource) error {
+	m := &migrationv1alpha1.StorageVersionMigration{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: storageStateName(resource),
+		},
+		Spec: migrationv1alpha1.StorageVersionMigrationSpec{
+			Resource: resource,
+		},
+	}
+	_, err := mt.client.MigrationV1alpha1().StorageVersionMigrations(namespaceName).Create(m)
+	return err
+}
+
+func (mt *MigrationTrigger) processMigration(m *migrationv1alpha1.StorageVersionMigration) error {
+	switch {
+	case controller.HasCondition(m, migrationv1alpha1.MigrationSucceeded):
+		return mt.markStorageStateSucceeded(m.Spec.Resource)
+	case controller.HasCondition(m, migrationv1alpha1.MigrationFailed):
```

Comment:
wondering if there is difference between permanent failure vs transient failure. for permanent failure, probably the best is to not retry, instead, having the operator to monitor for it, and take action. for transient failure, should retry have some backoff?

Reply:
Updated to do nothing in case of migration failed and documented the reasoning.

I'll add metrics for failed migrations in future PRs.